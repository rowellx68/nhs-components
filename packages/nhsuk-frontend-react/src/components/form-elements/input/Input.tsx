import React, { ReactNode } from 'react';
import {
  BaseFormElementProps,
  FormGroup,
} from '@/components/core/form-group/FormGroup';
import { Factory, factory } from '@/internal/factory/factory';
import { ColumnWidth, ElementProps, InputWidth } from '@/types/shared';
import clsx from 'clsx';
import { Base } from '@/components/core/base/Base';

export type InputProps = {
  width?: InputWidth | ColumnWidth;
  prefix?: ReactNode;
  suffix?: ReactNode;
} & BaseFormElementProps &
  ElementProps<'input', 'size' | 'as'>;

type InputFactory = Factory<{
  props: InputProps;
  ref: HTMLInputElement;
}>;

const Input = factory<InputFactory>(
  ({ width, suffix, prefix, ...props }, ref) => {
    const component = suffix || prefix ? 'div' : React.Fragment;
    const baseProps =
      component === 'div' ? { className: 'nhsuk-input__wrapper' } : {};

    const widthNumberLike = width && !isNaN(parseInt(width, 10));

    return (
      <FormGroup
        as="input"
        withErrorLine
        {...props}
        ref={ref}
        inputType="input"
        render={({ id, name, className, withError, ...rest }) => (
          <Base<any> as={component} {...baseProps}>
            {prefix && (
              <div className="nhsuk-input__prefix" aria-hidden="true">
                {prefix}
              </div>
            )}
            <input
              id={id}
              name={name}
              className={clsx(
                'nhsuk-input',
                {
                  [`nhsuk-input--width-${width}`]: widthNumberLike,
                  [`nhsuk-u-width-${width}`]: !widthNumberLike && width,
                  'nhsuk-u-width-full': !widthNumberLike && !width,
                  'nhsuk-input--error': withError,
                },
                className,
              )}
              {...rest}
            />
            {suffix && (
              <div className="nhsuk-input__suffix" aria-hidden="true">
                {suffix}
              </div>
            )}
          </Base>
        )}
      />
    );
  },
);

Input.displayName = 'Input';

export { Input };
