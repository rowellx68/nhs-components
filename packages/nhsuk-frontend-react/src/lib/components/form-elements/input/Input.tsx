import FormGroup from '@/components/core/form-group/FormGroup'
import { FormElementProps } from '@/types/form-element-types'
import { InputWidth } from '@/types/nhsuk-sizes'
import clsx from 'clsx'
import {
  ForwardRefRenderFunction,
  HTMLInputTypeAttribute,
  HTMLProps,
  ReactNode,
  forwardRef,
} from 'react'

type InputProps = {
  width?: InputWidth
  type?: HTMLInputTypeAttribute
  prefix?: ReactNode
  suffix?: ReactNode
} & Omit<HTMLProps<HTMLInputElement>, 'type'> &
  FormElementProps

/**
 * Use text input to let users enter a single line of text.
 *
 * For more information on when to use this component, go to the [NHS Service Manuals website](https://service-manual.nhs.uk/design-system/components/text-input).
 *
 * @example
 * ```tsx
 * <Input id="example" label="Example" />
 * ```
 */
const Input: ForwardRefRenderFunction<HTMLInputElement, InputProps> = (
  props,
  ref,
): JSX.Element => {
  return (
    <FormGroup<InputProps>
      {...props}
      inputType="input"
      render={({ id, name, className, error, width, type = 'text', ...rest }) =>
        props.prefix || props.suffix ? (
          <div className="nhsuk-input__wrapper">
            {props.prefix && (
              <div className="nhsuk-input__prefix" aria-hidden="true">{props.prefix}</div>
            )}
            <input
              id={id}
              name={name}
              className={clsx(
                'nhsuk-input',
                {
                  [`nhsuk-input--width-${width}`]: width,
                  'nhsuk-input--error': error,
                },
                className,
              )}
              type={type}
              ref={ref}
              {...rest}
            />
            {props.suffix && (
              <div className="nhsuk-input__suffix" aria-hidden="true">{props.suffix}</div>
            )}
          </div>
        ) : (
          <input
            id={id}
            name={name}
            className={clsx(
              'nhsuk-input',
              {
                [`nhsuk-input--width-${width}`]: width,
                'nhsuk-input--error': error,
              },
              className,
            )}
            type={type}
            ref={ref}
            {...rest}
          />
        )
      }
    />
  )
}

const InputForwardRef = forwardRef(Input)

InputForwardRef.displayName = 'Input'

export default InputForwardRef
