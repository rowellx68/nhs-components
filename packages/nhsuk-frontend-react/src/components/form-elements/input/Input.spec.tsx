import React from 'react';
import { it, expect } from 'vitest';
import { render } from '@testing-library/react';
import { Input, InputProps } from './Input';

it('should render the input component', () => {
  const { container } = render(<Input label="First name" />);

  expect(container).toMatchSnapshot();
});

it('should render the input component with hint', () => {
  const { container } = render(
    <Input label="First name" hint="Enter your first name" />,
  );

  expect(container).toMatchSnapshot();
});

it('should render the input component with error', () => {
  const { container } = render(
    <Input label="First name" error="Error message goes here" />,
  );

  expect(container).toMatchSnapshot();
});

it.each([
  '2',
  '3',
  '4',
  '5',
  '10',
  '20',
  'full',
  'one-half',
  'one-quarter',
  'one-third',
  'three-quarters',
  'two-thirds',
] as InputProps['width'][])(
  'should render the input component with width %s',
  (width) => {
    const { container } = render(<Input label="First name" width={width} />);

    expect(container).toMatchSnapshot();
  },
);

it('should render the input with prefix', () => {
  const { container } = render(<Input label="Cost" prefix="Â£" />);

  expect(container).toMatchSnapshot();
});

it('should render the input with suffix', () => {
  const { container } = render(<Input label="Weight" suffix="in grams" />);

  expect(container).toMatchSnapshot();
});
