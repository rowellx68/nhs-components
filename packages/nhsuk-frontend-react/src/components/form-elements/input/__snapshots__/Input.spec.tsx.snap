// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render the input component 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:r0:"
      id="input_:r0:--label"
    >
      First name
    </label>
    <input
      aria-labelledby="input_:r0:--label"
      as="input"
      class="nhsuk-input nhsuk-u-width-full"
      id="input_:r0:"
    />
  </div>
</div>
`;

exports[`should render the input component with error 1`] = `
<div>
  <div
    class="nhsuk-form-group nhsuk-form-group--error"
  >
    <label
      class="nhsuk-label"
      for="input_:r4:"
      id="input_:r4:--label"
    >
      Full name
    </label>
    <p
      class="nhsuk-error-message"
      id="input_:r4:--error-message"
    >
      <span
        class="nhsuk-u-visually-hidden"
      >
        Error: 
      </span>
      Enter your full name
    </p>
    <input
      aria-describedby="input_:r4:--error-message"
      aria-labelledby="input_:r4:--label"
      as="input"
      class="nhsuk-input nhsuk-u-width-full nhsuk-input--error"
      id="input_:r4:"
    />
  </div>
</div>
`;

exports[`should render the input component with hint 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:r2:"
      id="input_:r2:--label"
    >
      Enter a full postcode in England
    </label>
    <div
      class="nhsuk-hint"
      id="input_:r2:_:r3:--hint"
    >
      For example, LS1 1AB
    </div>
    <input
      aria-describedby="input_:r2:_:r3:--hint"
      aria-labelledby="input_:r2:--label"
      as="input"
      class="nhsuk-input nhsuk-input--width-10"
      id="input_:r2:"
    />
  </div>
</div>
`;

exports[`should render the input component with width 2 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:r6:"
      id="input_:r6:--label"
    >
      First name
    </label>
    <input
      aria-labelledby="input_:r6:--label"
      as="input"
      class="nhsuk-input nhsuk-input--width-2"
      id="input_:r6:"
    />
  </div>
</div>
`;

exports[`should render the input component with width 3 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:r8:"
      id="input_:r8:--label"
    >
      First name
    </label>
    <input
      aria-labelledby="input_:r8:--label"
      as="input"
      class="nhsuk-input nhsuk-input--width-3"
      id="input_:r8:"
    />
  </div>
</div>
`;

exports[`should render the input component with width 4 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:ra:"
      id="input_:ra:--label"
    >
      First name
    </label>
    <input
      aria-labelledby="input_:ra:--label"
      as="input"
      class="nhsuk-input nhsuk-input--width-4"
      id="input_:ra:"
    />
  </div>
</div>
`;

exports[`should render the input component with width 5 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:rc:"
      id="input_:rc:--label"
    >
      First name
    </label>
    <input
      aria-labelledby="input_:rc:--label"
      as="input"
      class="nhsuk-input nhsuk-input--width-5"
      id="input_:rc:"
    />
  </div>
</div>
`;

exports[`should render the input component with width 10 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:re:"
      id="input_:re:--label"
    >
      First name
    </label>
    <input
      aria-labelledby="input_:re:--label"
      as="input"
      class="nhsuk-input nhsuk-input--width-10"
      id="input_:re:"
    />
  </div>
</div>
`;

exports[`should render the input component with width 20 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:rg:"
      id="input_:rg:--label"
    >
      First name
    </label>
    <input
      aria-labelledby="input_:rg:--label"
      as="input"
      class="nhsuk-input nhsuk-input--width-20"
      id="input_:rg:"
    />
  </div>
</div>
`;

exports[`should render the input component with width full 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:ri:"
      id="input_:ri:--label"
    >
      First name
    </label>
    <input
      aria-labelledby="input_:ri:--label"
      as="input"
      class="nhsuk-input"
      id="input_:ri:"
    />
  </div>
</div>
`;

exports[`should render the input component with width one-half 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:rk:"
      id="input_:rk:--label"
    >
      First name
    </label>
    <input
      aria-labelledby="input_:rk:--label"
      as="input"
      class="nhsuk-input nhsuk-u-width-one-half"
      id="input_:rk:"
    />
  </div>
</div>
`;

exports[`should render the input component with width one-quarter 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:rm:"
      id="input_:rm:--label"
    >
      First name
    </label>
    <input
      aria-labelledby="input_:rm:--label"
      as="input"
      class="nhsuk-input nhsuk-u-width-one-quarter"
      id="input_:rm:"
    />
  </div>
</div>
`;

exports[`should render the input component with width one-third 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:ro:"
      id="input_:ro:--label"
    >
      First name
    </label>
    <input
      aria-labelledby="input_:ro:--label"
      as="input"
      class="nhsuk-input nhsuk-u-width-one-third"
      id="input_:ro:"
    />
  </div>
</div>
`;

exports[`should render the input component with width three-quarters 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:rq:"
      id="input_:rq:--label"
    >
      First name
    </label>
    <input
      aria-labelledby="input_:rq:--label"
      as="input"
      class="nhsuk-input nhsuk-u-width-three-quarters"
      id="input_:rq:"
    />
  </div>
</div>
`;

exports[`should render the input component with width two-thirds 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:rs:"
      id="input_:rs:--label"
    >
      First name
    </label>
    <input
      aria-labelledby="input_:rs:--label"
      as="input"
      class="nhsuk-input nhsuk-u-width-two-thirds"
      id="input_:rs:"
    />
  </div>
</div>
`;

exports[`should render the input with prefix 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:ru:"
      id="input_:ru:--label"
    >
      Cost in pounds
    </label>
    <div
      class="nhsuk-input__wrapper"
    >
      <div
        aria-hidden="true"
        class="nhsuk-input__prefix"
      >
        £
      </div>
      <input
        aria-labelledby="input_:ru:--label"
        as="input"
        class="nhsuk-input nhsuk-input--width-5"
        id="input_:ru:"
      />
    </div>
  </div>
</div>
`;

exports[`should render the input with prefix and suffix 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:r12:"
      id="input_:r12:--label"
    >
      Cost per item, in pounds
    </label>
    <div
      class="nhsuk-input__wrapper"
    >
      <div
        aria-hidden="true"
        class="nhsuk-input__prefix"
      >
        £
      </div>
      <input
        aria-labelledby="input_:r12:--label"
        as="input"
        class="nhsuk-input nhsuk-input--width-5"
        id="input_:r12:"
      />
      <div
        aria-hidden="true"
        class="nhsuk-input__suffix"
      >
        per item
      </div>
    </div>
  </div>
</div>
`;

exports[`should render the input with prefix, suffix and error 1`] = `
<div>
  <div
    class="nhsuk-form-group nhsuk-form-group--error"
  >
    <label
      class="nhsuk-label"
      for="input_:r14:"
      id="input_:r14:--label"
    >
      Cost per item, in pounds
    </label>
    <p
      class="nhsuk-error-message"
      id="input_:r14:--error-message"
    >
      <span
        class="nhsuk-u-visually-hidden"
      >
        Error: 
      </span>
      Enter a cost per item, in pounds
    </p>
    <div
      class="nhsuk-input__wrapper"
    >
      <div
        aria-hidden="true"
        class="nhsuk-input__prefix"
      >
        £
      </div>
      <input
        aria-describedby="input_:r14:--error-message"
        aria-labelledby="input_:r14:--label"
        as="input"
        class="nhsuk-input nhsuk-input--width-5 nhsuk-input--error"
        id="input_:r14:"
      />
      <div
        aria-hidden="true"
        class="nhsuk-input__suffix"
      >
        per item
      </div>
    </div>
  </div>
</div>
`;

exports[`should render the input with suffix 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:r10:"
      id="input_:r10:--label"
    >
      Weight in kilograms
    </label>
    <div
      class="nhsuk-input__wrapper"
    >
      <input
        aria-labelledby="input_:r10:--label"
        as="input"
        class="nhsuk-input nhsuk-input--width-5"
        id="input_:r10:"
      />
      <div
        aria-hidden="true"
        class="nhsuk-input__suffix"
      >
        kg
      </div>
    </div>
  </div>
</div>
`;

exports[`should render the input with whole numbers 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label"
      for="input_:r16:"
      id="input_:r16:--label"
    >
      What is your NHS number?
    </label>
    <div
      class="nhsuk-hint"
      id="input_:r16:_:r17:--hint"
    >
      Your NHS number is a 10 digit number that you find on any letter the NHS has sent you. For example, 485 777 3456.
    </div>
    <input
      aria-describedby="input_:r16:_:r17:--hint"
      aria-labelledby="input_:r16:--label"
      as="input"
      class="nhsuk-input nhsuk-input--width-10"
      id="input_:r16:"
      inputmode="numeric"
      pattern="[0-9]*"
      type="text"
    />
  </div>
</div>
`;
